aws_lambda_powertools/utilities/data_classes/kinesis_firehose_event.py

```diff
@@ -1,9 +1,125 @@
 import base64
-from typing import Iterator, Optional
+import json
+from dataclasses import dataclass, field
+from typing import Any, Callable, Dict, Iterator, List, Optional
+
+from typing_extensions import Literal
 
 from aws_lambda_powertools.utilities.data_classes.common import DictWrapper
 
 
+@dataclass(repr=False, order=False, frozen=True)
+class KinesisFirehoseDataTransformationRecordMetadata:
+    """
+    Documentation:
+    --------------
+    - https://docs.aws.amazon.com/firehose/latest/dev/dynamic-partitioning.html
+    """
+
+    partition_keys: Optional[Dict[str, str]]
+
+    def asdict(self) -> Dict:
+        if self.partition_keys is not None:
+            return {"partitionKeys": self.partition_keys}
+        return {}
+
+
+@dataclass(repr=False, order=False)
+class KinesisFirehoseDataTransformationRecord:
+    """Record in Kinesis Data Firehose response object
+
+    Documentation:
+    --------------
+    - https://docs.aws.amazon.com/firehose/latest/dev/data-transformation.html
+    """
+
+    # Record ID; uniquely identifies this record within the current batch"""
+    record_id: str
+    # Processing result, supported value: Ok, Dropped, ProcessingFailed"""
+    result: Literal["Ok", "Dropped", "ProcessingFailed"] = "Ok"
+    # data blob, base64-encoded, optional at init. Allows pass in base64-encoded data directly or
+    # use either function like `data_from_text`, `data_from_json` to populate data"""
+    data: Optional[str] = None
+    # Optional: Metadata associated with this record; can contain partition keys
+    # See - https://docs.aws.amazon.com/firehose/latest/dev/dynamic-partitioning.html
+    metadata: Optional[KinesisFirehoseDataTransformationRecordMetadata] = None
+    _json_data: Optional[Any] = None
+    json_serializer: Callable = json.dumps
+    json_deserializer: Callable = json.loads
+
+    def data_from_byte(self, data: bytes):
+        """Populate data field using a byte like data"""
+        self.data = base64.b64encode(data).decode("utf-8")
+
+    def data_from_text(self, data: str):
+        """Populate data field using a string like data"""
+        self.data_from_byte(data.encode("utf-8"))
+
+    def data_from_json(self, data: Any):
+        """Populate data field using any structure that could be converted to json"""
+        self.data_from_text(data=self.json_serializer(data))
```

Comment:
```suggestion
```

In-place mutation can almost always go wrong. It looks like what you're trying to improve in UX is make it easier to transform data that will be added in the final processed_record.

Since you accept a `data=` argument, it's easier to have a bunch of standalone functions that handle X to base64.

---

Example

```python
        processed_record = KinesisFirehoseDataTransformationRecord(
            record_id=record.record_id, result="Ok", metadata=metadata_partition, data=base64_from_json(data)
        )
        response.add_record(record=processed_record)
```